### PR TITLE
Use proxy for every request, not only for posting data.

### PR DESCRIPTION
**Pull request by lemavri**
This PR removes the logic related to proxy from `lib/agent/plugins/control-panel/sender.js` to `lib/agent/plugins/control-panel/api/request.js`. This way, it ensures all requests are sent through a proxy if set in the config file.

### Implementation

If try_proxy is defined in the config file, its value is set as default during the api initialization. 

When sending a request with an existing `options.proxy`, there are a few different use cases:
- If the request fails due to a network error (in this case it's specifically `EHOSTUNREACH `), then it retries without proxy
  - If the retry without proxy fails due to network error, the request is dismissed
  - If the retry without proxy fails due to temp error, retries 2 more times. If all retries fail then the request is dismissed
- If the request fails due to a temp error:
  - And `opts.proxy` is set, it retries 2 more times using proxy. If all the retries fail, fallbacks to no proxy.

BTW, I had to add the function `set_proxy_and_send` because I found this to be the best way of assigning the `opts.proxy` and `options.proxy` values.

_**bonus:** semicolon invasion_

Log files:

_Observation: Due to the async nature of the requests, the following logs might not be in logical order._

### Proxy set but host does not exist
Fallback to non proxy request right away.
#### Fetching instructions
screen shot 2015-03-04 at 12 52 39 pm

#### Posting data
screen shot 2015-03-04 at 12 41 57 pm

### Proxy set but refusing connection (temp error)
Fallback to non proxy request after 3 retries.
#### Fetching instructions
screen shot 2015-03-04 at 2 04 26 pm

#### Posting data
screen shot 2015-03-04 at 2 05 27 pm

### Proxy set and reachable
Use the proxy.
#### Fetching instructions
screen shot 2015-03-04 at 1 22 09 pm

#### Posting data
Obs: `401` response is due to [needle](https://github.com/tomas/needle/blob/master/lib/needle.js#L126-L133) not yet passing the credentials as basic auth to the destination.
screen shot 2015-03-04 at 1 24 22 pm

### CODE REVIEW

File: lib/agent/plugins/control-panel/api/request.js [hunk 1]
```diff
@@ -53,18 +57,30 @@ var send = function(attempt, method, path, data, options, cb) {
...
+          logger.debug('Retrying request without proxy.');
+          send(1, method, path, data, options, cb);
+        };
     logger.debug('Attempt #' + attempt + ' took ' + seconds + ' seconds.');
 
     if (err && is_network_down(err)) {
 
       err.message = 'Network seems to be down. Check your connection and try again.';
+      if(opts.proxy) {
+        retry_without_proxy();
```
Comment (tomas) [inline lib/agent/plugins/control-panel/api/request.js]: Isn't there a missing `return` statement right here?

Comment (lemavri) [inline lib/agent/plugins/control-panel/api/request.js]: Yep, I think you're right. In that case line 94 also needs a return, right?

**Comment by tomas** [comment]
Looks good. We need to take a look at the init_api() call in `lib/conf/shared/panel` as well, though. 

Question: Why retry all requests without proxy on temporary failures? If the request went through and we got a 503 then retrying it without a proxy won't make any difference -- actually, the most probable outcome will be a network error (unable to connect).

**Comment by lemavri** [comment]
That's a good question, that I actually forgot to document. 

According to my tests, there are two kind of responsed that might be causes by proxy failure:

is_network_down (`EHOSTUNREACH`): `try_proxy` is set but the host does not exist.
is_temporary_error: `try_proxy` is set, host is resolving, but is refusing the connection, which might be a temporary error of the proxy itself (overload or some other reason). That's why I think it's a good idea to retry the connection using the proxy and then fallback.

**Comment by lemavri** [comment]
Added the missing return statements. Without these the error trigger was jumping even if the connection was finally successful.

**Comment by lemavri** [comment]
> We need to take a look at the init_api() call in lib/conf/shared/panel as well, though.

I'm lost, what should be done here?

**Comment by tomas** [comment]
The API initialization done as part of the conf module: https://github.com/prey/prey-node-client/blob/master/lib/conf/shared/panel.js#L21

**Comment by lemavri** [comment]
Got it, so try_proxy should be appended to opts as well, amirite?

**Comment by lemavri** [comment]
Done, but I need the needle fix to test it with the proxy enabled (I'm getting `Invalid credentials` since needle does not pass the basic auth to control).

**Comment by lemavri** [comment]
This is the result of a POST request using a proxy and needle rc (no longer `401` response).

screen shot 2015-03-04 at 4 30 39 pm

**Comment by lemavri** [comment]
These are the refactored specs:

screen shot 2015-03-06 at 12 35 43 pm

**Comment by tomas** [comment]
Beautiful. I say go!

**Comment by lemavri** [comment]
And off we go 👊

**Comment by fanuneza** [comment]
Buenaaaaaaaa :+1:

**Pull request merged**